lib/courses/bloc.dart

```diff
@@ -0,0 +1,49 @@
+import 'package:flutter/foundation.dart';
+import 'package:rxdart/subjects.dart';
+import 'package:schulcloud/app/services.dart';
+import 'package:schulcloud/core/data.dart';
+import 'package:schulcloud/core/data/utils.dart';
+import 'package:schulcloud/courses/data/course.dart';
+import 'package:schulcloud/courses/data/lesson.dart';
+import 'package:schulcloud/courses/data/repository.dart';
+
+import 'entities.dart';
+
+export 'entities.dart';
+
+class Bloc {
+  final ApiService api;
+  Repository<Course> _courses;
+
+  Bloc({@required this.api})
+      : _courses = CachedRepository<Course>(
+          source: CourseDownloader(api: api),
+          cache: InMemoryStorage<Course>(),
+        );
+
+  Stream<List<Course>> getCourses() {
+    return streamToBehaviorSubject(_courses.fetchAllItems());
+  }
+
+  BehaviorSubject<Course> getCourseAtIndex(int index) {
+    final BehaviorSubject<Course> s =
+        streamToBehaviorSubject(_courses.fetch(Id('course_$index')));
+    s.listen((data) {
+      print(data);
+    });
+    return s;
+  }
+
+  Stream<List<Lesson>> getLessons(Id<Course> courseId) {
+    var _lessons = CachedRepository<Lesson>(
```

Comment:
Probably makes sense to define a `_lessons` repository at the top of the class and re-use it instead of creating a new one every time the widget layer fetches lessons

Reply:
Considering that the lesson repository has to be created for each course individually, creating these repositories on the fly is (for the time being) the simplest way to do this. This may be changed at a later time, but for now it will be left as it is.